pkg/controller/machinepool/machinepool_controller.go

```diff
@@ -513,7 +694,141 @@ func (r *ReconcileRemoteMachineSet) ensureEnoughReplicas(
 	return nil, nil
 }
 
-func (r *ReconcileRemoteMachineSet) syncMachineSets(
+func (r *ReconcileMachinePool) ensureEnoughReplicasCAPI(
+	pool *hivev1.MachinePool,
+	generatedMachineSets []*capiv1.MachineSet,
+	cd *hivev1.ClusterDeployment,
+	logger log.FieldLogger,
+) (*reconcile.Result, error) {
+	if pool.Spec.Autoscaling == nil {
+		return nil, nil
+	}
+	if pool.Spec.Autoscaling.MinReplicas < int32(len(generatedMachineSets)) && !platformAllowsZeroAutoscalingMinReplicas(cd) {
+		logger.WithField("machinesets", len(generatedMachineSets)).
+			WithField("minReplicas", pool.Spec.Autoscaling.MinReplicas).
+			Warning("when auto-scaling, the MachinePool must have at least one replica for each MachineSet")
+		conds, changed := controllerutils.SetMachinePoolConditionWithChangeCheck(
+			pool.Status.Conditions,
+			hivev1.NotEnoughReplicasMachinePoolCondition,
+			corev1.ConditionTrue,
+			"MinReplicasTooSmall",
+			fmt.Sprintf("When auto-scaling, the MachinePool must have at least one replica for each MachineSet. The minReplicas must be at least %d", len(generatedMachineSets)),
+			controllerutils.UpdateConditionIfReasonOrMessageChange,
+		)
+		if changed {
+			pool.Status.Conditions = conds
+			if err := r.Status().Update(context.Background(), pool); err != nil {
+				logger.WithError(err).Error("failed to update MachinePool conditions")
+				return &reconcile.Result{}, err
+			}
+		}
+		return &reconcile.Result{}, nil
+	}
+	conds, changed := controllerutils.SetMachinePoolConditionWithChangeCheck(
+		pool.Status.Conditions,
+		hivev1.NotEnoughReplicasMachinePoolCondition,
+		corev1.ConditionFalse,
+		"EnoughReplicas",
+		"The MachinePool has sufficient replicas for each MachineSet",
+		controllerutils.UpdateConditionNever,
+	)
+	if changed {
+		pool.Status.Conditions = conds
+		err := r.Status().Update(context.Background(), pool)
+		if err != nil {
+			logger.WithError(err).Error("failed to update MachinePool conditions")
+		}
+		return &reconcile.Result{}, err
+	}
+	return nil, nil
+}
+
+func (r *ReconcileMachinePool) syncAWSMachineTemplates(
```

Comment:
my thinking is,
1. the machine template types do not have any common structure are very actuator specific.
2. the only enforcement that is possible from high level of controller(reconciler) is the k8s object metadata.
3. So the general implementation of the sync of machine templates should act on client.Object ensuring metadata or cleanup. anything more would require actuator help.

And i think it's good enough to begin with.. since even for sync machinesets we only take the machineapi common definitions into account.